src/aruku/walking/process/kinematic.cpp

```diff
@@ -0,0 +1,729 @@
+  m_z_move_amplitude = 0.0;
+
+  update_times();
+  update_move_amplitude();
+}
+
+double Kinematic::wsin(
+  double time, double period, double period_shift, double mag,
+  double mag_shift) const
+{
+  return mag * sin(2_pi / period * time - period_shift) + mag_shift;
+}
+
+bool Kinematic::compute_inverse_kinematic(
+  int index_addition, keisan::Point3 translation_target, keisan::Euler<double> rotation_target)
+{
+  using keisan::Point3;
+  using keisan::Euler;
+  using keisan::Matrix;
+  using tachimawari::joint::JointId;
+
+  translation_target.z -= leg_length;
+  auto matrix_translation = keisan::translation_matrix(translation_target);
+  auto matrix_rotation = keisan::rotation_matrix(rotation_target);
+
+  auto matrix_transformation = matrix_translation * matrix_rotation;
+
+  auto vector = Point3(
+    translation_target.x + (matrix_transformation[0][2] * ankle_length),
+    translation_target.y + (matrix_transformation[1][2] * ankle_length),
+    translation_target.z + (matrix_transformation[2][2] * ankle_length));
+
+  // get knee
+  double vector_magnitude = vector.magnitude();
+  double acos_result = acos(
+    ((vector_magnitude * vector_magnitude) - (thigh_length * thigh_length) -
+    (calf_length * calf_length)) / (2 * thigh_length * calf_length));
+
+  if (std::isnan(acos_result)) {
+    return false;
+  }
+  angles[static_cast<size_t>(JointId::RIGHT_KNEE) + index_addition] = acos_result;
+
+  // get ankle roll
+  auto matrix = matrix_transformation;
+  if (!matrix.inverse()) {
+    return false;
+  }
+
+  double k = sqrt(matrix[1][3] * matrix[1][3] + matrix[2][3] * matrix[2][3]);
+  double l =
+    sqrt(
+    matrix[1][3] * matrix[1][3] + (matrix[2][3] - ankle_length) *
+    (matrix[2][3] - ankle_length));
+  double m = (k * k - l * l - ankle_length * ankle_length) / (2 * l * ankle_length);
+  m = keisan::clamp(m, -1.0, 1.0);
+
+  acos_result = acos(m);
+  if (std::isnan(acos_result)) {
+    return false;
+  }
+  if (matrix[1][3] < 0.0) {
+    angles[JointId::RIGHT_ANKLE_PITCH + index_addition] = -acos_result;
+  } else {
+    angles[JointId::RIGHT_ANKLE_PITCH + index_addition] = acos_result;
+  }
+
+  // get hip yaw
+  matrix_translation = keisan::translation_matrix(Point3(0, 0, ankle_length));
+  matrix_rotation =
+    keisan::rotation_matrix(
+    Euler(
+      keisan::make_radian(angles[JointId::RIGHT_ANKLE_PITCH + index_addition]), 0.0_pi_rad,
+      0.0_pi_rad));
+
+  matrix = matrix_translation * matrix_rotation;
+  if (!matrix.inverse()) {
+    return false;
+  }
+
+  matrix = matrix_transformation * matrix;
+  double atan_result = atan2(-matrix[0][1], matrix[1][1]);
+  if (std::isinf(atan_result)) {
+    return false;
+  }
+  angles[JointId::RIGHT_HIP_YAW + index_addition] = atan_result;
+
+  // get hip roll
+  atan_result =
+    atan2(
+    matrix[2][1],
+    -matrix[0][1] *
+    sin(
+      angles[JointId::RIGHT_HIP_YAW + index_addition] + matrix[1][1] *
+      cos(angles[JointId::RIGHT_HIP_YAW + index_addition])));
```

Comment:
please separate this into several declaration, instead of using this format, e.g.:
```
auto a = b * c + d * e;
// to
auto bc = b * c;
auto de = d * e;
auto a = bc + de;
```
check other lines too

Reply:
done